#API-4731 production access request emails

reedship opened the pull request:
This PR closes [#API-4731](https://vajira.max.gov/secure/RapidBoard.jspa?rapidView=1185&projectKey=API&view=detail&selectedIssue=API-4731).

The purpose of this additional route is to allow for production access requests to be forwarded to support in a readable way, and for consumers to receive an email confirming their request and describing future steps.

Integration and unit tests have been added. The request body schema may change in the future, but this iteration is the initial best guess of how the request form data will be passed, as it is not yet implemented. Should the schema change, tests and validators can be updated for the new schema.

The email template design for consumers can be seen [here](https://www.figma.com/file/9sOD2GDAsyNTayvOqKlcJV/API-3854-Consumer?node-id=1037%3A6740). You can see the email generated for consumers below. This differs slightly from the design on the margins of the `<p>` tags below Steps 2-4 but it's as close as I can get the template once it goes through whatever voodoo the email client does to passed-in html.

Inside email:
![Screen Shot 2021-07-13 at 2 37 03 PM](https://user-images.githubusercontent.com/27251568/125514430-0c3f2d58-8505-4b47-90b7-56acdf98456e.png)

Base HTML:
![Screen Shot 2021-07-01 at 3 00 44 PM](https://user-images.githubusercontent.com/27251568/124183012-3e8b0f00-da7d-11eb-87dd-f751fd5119f9.png)

The email sent to support is just a form data dump. See below:
[Oddball Mail - Production Access Requested for Sauron.INC.pdf](https://github.com/department-of-veterans-affairs/developer-portal-backend/files/6811495/Oddball.Mail.-.Production.Access.Requested.for.Sauron.INC.pdf)

reedship committed:
added base production request support template. consumer facing coming soon

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Updated contactUsHandler to account for new type.

necessary types and interfaces have been added. This will require a
strict `req.body` schema, but it's managable.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
working commit: consumer facing production access template

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
updated links to production access consumer template

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Merge branch 'master' into api-4731

reedship committed:
added route, handler, and test for govdelivery method

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
working commit for ProductionRequest.test.ts

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Merge branch 'master' into api-4731

reedship committed:
added test to cover send_email functionality

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
removed dead code from ContactUs.ts

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
email functionality is working now. template updated for readability.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
All functionality and tests complete

Emails are being sent as expected to the statusUpdateEmails provided in
req.body and to the SUPPORT_EMAIL defined in environment variables.
Img sources for the production consumer template are the public object URLs and are fine
to be hard coded in the HTML element. An integration test for the route
has been added with this commit as well. All tests are passing and
running `npm run lint` returns no errors.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
removed dead code from contactus

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Merge branch 'master' into api-4731-production-access-request-emails

reedship committed:
adjusted margin in consumer email template

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Merge branch 'api-4731-production-access-request-emails' of github.com:department-of-veterans-affairs/developer-portal-backend into api-4731-production-access-request-emails

SlakrHakr commented on routes/ProductionRequest.integration.ts:
```diff
@@ -0,0 +1,132 @@
+    signUpLink: 'www.one2bindthem.com/signup',
+    supportLink: 'www.one2bindthem.com/support',
+    platforms: ['iOS'],
+    veteranFacingDescription: 'Now the Elves made many rings; but secretly Sauron made One Ring to rule all the others, and their power was bound up with it, to be subject wholly to it and to last only so long as it too should last.',
```
![gandalf-da-wizard](https://media.giphy.com/media/Qs75BqLW44RrP0x6qL/giphy.gif)

reedship committed:
Merge branch 'master' into api-4731-production-access-request-emails

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  vasiSystemName: Joi.string(),
+  credentialStorage: Joi.string().required(),
+  storePIIOrPHI: Joi.boolean().required(),
+  storageMethod: Joi.string(),
+  safeguards: Joi.string(),
```
i think these could use a bit more descriptive names. "piiStorageMethod", "multipleReqSafeguards"

bradenshipley replied:
will be in next commit

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  safeguards: Joi.string(),
+  breachManagementProcess: Joi.string(),
+  vulnerabilityManagement: Joi.string(),
+  exposeHealthInformationToThirdParties: Joi.boolean(),
+  thirdPartyHealthInfoDescription: Joi.string(),
```
i think it includes other areas as well (claims, service history, etc)?

```suggestion
  exposeVeteranInformationToThirdParties: Joi.boolean(),
  thirdPartyInfoDescription: Joi.string(),
```

bradenshipley replied:
will be in next commit

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  exposeHealthInformationToThirdParties: Joi.boolean(),
+  thirdPartyHealthInfoDescription: Joi.string(),
+  scopesAccessRequested: Joi.array().items(Joi.string()),
+  distrubitingAPIKeysToCustomers: Joi.boolean(),
```
```suggestion
  distributingAPIKeysToCustomers: Joi.boolean(),
```

bradenshipley replied:
ah nice typo catch. will be in next commit

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  thirdPartyHealthInfoDescription: Joi.string(),
+  scopesAccessRequested: Joi.array().items(Joi.string()),
+  distrubitingAPIKeysToCustomers: Joi.boolean(),
+  namingConvention: Joi.string(),
+  centralizedBackendLog: Joi.string(),
```
would clarify these are only for Benefits Intake consumers that are allowing their customers to run hosted instances of the app. also think we're missing the question that prompts these:

> Will you be distributing your production API key to any customers for them to run their own hosted instances of your application? (*Required)

bradenshipley replied:
that question is covered by this line here: 
`  distributingAPIKeysToCustomers?: boolean;`

bradenshipley replied:
also needs a `required()`. will be in next commit

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  namingConvention: Joi.string(),
+  centralizedBackendLog: Joi.string(),
+  listedOnMyHealthApplication: Joi.boolean(),
+}).options({ abortEarly: false });
```
missing links to terms of service pages?

bradenshipley replied:
held inside `policyDocuments` https://github.com/department-of-veterans-affairs/developer-portal-backend/blob/4b4ffa06a3b2983cd18be75e1ad1b369b75ba8e7/routes/ProductionRequest.ts#L25

mjlumetta commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+    next: NextFunction
+  ): Promise<void> {
+    try {
+      if (govdelivery) {
```
hate that this is still a thing but whatever, now's probably not the right time to worry about that

mjlumetta commented on services/GovDeliveryService.ts:
```diff
@@ -48,6 +52,53 @@ export interface PublishingSupportEmail {
+  lastName: string;
+  email: string;
+}
+export interface ProductionAccessSupportEmail {
```
i would love to move these to the `types` directory. don't have to migrate the old ones, but i think we can put the new ones there

bradenshipley replied:
that's a good idea. I'll move them

mjlumetta commented on services/GovDeliveryService.ts:
```diff
@@ -48,6 +52,53 @@ export interface PublishingSupportEmail {
+  centralizedBackendLog?: string;
+  listedOnMyHealthApplication?: boolean;
+}
+export interface ProductionAccessConsumerEmail {
```
is this actually used anywhere?

bradenshipley replied:
No this is dead code from when I was trying to send the images as base64. removing in next commit

mjlumetta commented on services/GovDeliveryService.ts:
```diff
@@ -173,6 +228,29 @@ export default class GovDeliveryService implements MonitoredService {
 
+  public sendProductionAccessEmail(supportRequest: ProductionAccessSupportEmail): Promise<EmailResponse> {
+    const email: EmailRequest = {
+      subject: 'Production Access Requested',
```
might be helpful to include the organization name in the subject – guessing that'll make it easier to sort through these in the inbox

bradenshipley replied:
added in latest commit

mjlumetta commented on templates/index.ts:
```diff
@@ -1,131 +1,131 @@
-    </div>
-</li>
-</ul>`;
+export const WELCOME_TEMPLATE = `<div>Welcome {{ firstName }},</div><br />
```
is there an actual diff in this file? if no can we maybe just check out the old version so it's less confusing? or we can use it as a barrel file and export the production access templates from here as well.

bradenshipley replied:
yeah i think this diff will be fixed by merging in the latest main branch

bradenshipley replied:
i do think breaking up the production templates into their own file is 💯 % the more readable option

mjlumetta commented on templates/production.ts:
```diff
@@ -0,0 +1,368 @@
+  </style>
+</head>
+<div class="card">
+  <div class="header"><img src="https://dvp-developer-portal-backend-images.s3-us-gov-west-1.amazonaws.com/VA-Lighhouse-APIs-email.png" alt="logo"/></div>
```
maybe make the alt "VA Lighthouse APIs" or "VA Lighthouse APIs logo"?

bradenshipley replied:
will be in next commit

mjlumetta commented on templates/production.ts:
```diff
@@ -0,0 +1,368 @@
+    production access takes a week or less and no demo is required.</p>
+  <h1><strong>What’s next?</strong></h1>
+  <div class="steps-container">
+    <img class="steps" src="https://dvp-developer-portal-backend-images.s3-us-gov-west-1.amazonaws.com/Done.png" alt="Step one, done"/><h2>Start developing.</h2>
```
these probably shouldn't be h2's because they're not really covering a section per se. also could maybe wrap these in lists rather than divs, since it is semantically a list

bradenshipley replied:
Aren't they each covering a section? They are individual sections of the `What's Next?` part of the template.

johanna-adhoc replied:
I would add the list structure but I would keep the divs. I would also just use the strong tag to apply the bold effect and replace the heading with <p> elements. Example:
<ul>
<li>
<div class="steps-container">
    <img class="steps" src="..." alt="Step two, done"/>
    <p><strong>Submit your production access request.</strong></p>
  </div>
  <p class="sub">
    We’ll review your information and notify you within 1-2 weeks if there are any changes we need.
  </p>
</li>
</ul>

The nice thing about the li is that it groups the step description and its details.

bradenshipley replied:
These have been changed to use a list in the latest commit!

mjlumetta commented on templates/production.ts:
```diff
@@ -0,0 +1,368 @@
+  </p>
+  <p>
+    <a href="https://developer.va.gov/onboarding/prepare-for-and-complete-a-demo">
+      <u>Preparing for a demo</u>
```
think the underline should be covered by the link style, not the element. also think `text-decoration: underline` is preferred if it needs to be styled that way

bradenshipley replied:
will be in next commit

mjlumetta commented on templates/production.ts:
```diff
@@ -0,0 +1,368 @@
+</div>
+`;
+export const PRODUCTION_ACCESS_SUPPORT_TEMPLATE = `<h2>Basic Information</h2>
+<ul>
```
personally don't think the `ul` format is super readable for this sort of lengthy information (i know we use it for other support emails), but as long as support can read it, i'm fine with it

bradenshipley replied:
I'd leave that decision up to support to decide later on. The format can always be updated.

maddy-jo reviewed (commented):
have not really tested but mostly this looks good. most of these comments are detail questions about the schema

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+import { DevPortalError } from '../models/DevPortalError';
+import {validateApiList} from './DeveloperApplication';
+
+export const productionSchema = Joi.object().keys({
```
On the form if the consumer answers `NO` to either of those questions, the requester should be notified that these are requirements, so my thought was thought it wasn't necessary to track those.
![Screen Shot 2021-07-08 at 9 35 49 AM](https://user-images.githubusercontent.com/27251568/124940787-e3eb3900-dfcf-11eb-9bd0-358bd596382b.png)

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+    email: Joi.string().required(),
+  }).required(),
+  organization: Joi.string().required(),
+  appName: Joi.string().required(),
```
I'm just now seeing the comment you made on the 'all form fields' doc. I had a similar thought and figured i'd follow the '(*Required)` tag until given further clarification.

bradenshipley commented on routes/DeveloperApplication.ts:
```diff
@@ -11,7 +11,7 @@ import DynamoService from '../services/DynamoService';
 import { DevPortalError } from '../models/DevPortalError';
 
-function validateApiList(val: string): string {
+export function validateApiList(val: string): string {
```
that's a good idea. I'll put it in `utils`.

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  valueProvided: Joi.string().required(),
+  businessModel: Joi.string(),
+  policyDocuments: Joi.array().items(Joi.string()).required(),
+  phoneNumber: Joi.string().required(),
```
I had assumed format validation for the input would be handled in the form. I'm fine with adding it to this though. What formats are we trying to cover? 
Example formats:
- (123) 456-7890
- (123)456-7890
- 123-456-7890
- 123.456.7890
- 1234567890

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  apis: Joi.custom(validateApiList).required(),
+  monitizedVeteranInformation: Joi.boolean().required(),
+  monitizationExplanation: Joi.string(),
+  veteranFacing: Joi.boolean(),
```
it is indeed. Will add in next commit.

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  businessModel: Joi.string(),
+  policyDocuments: Joi.array().items(Joi.string()).required(),
+  phoneNumber: Joi.string().required(),
+  apis: Joi.custom(validateApiList).required(),
```
I was considering it to be passed in as a comma delimited string since it's how the existing developer access request is handling requested apis being passed. The conditional validation already existed so it was an easy 'plug and play' decision. Making the validator handle either would be easy to do, and the array makes more sense it my head as well.

reedship committed:
Moved validateApiList to utils directory

also made veteranFacing required in productionRequest.ts
added email() validation to primaryContact.email and
secondaryContact.email fields.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
added max() validation for veteranFacingDescription

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  signUpLink: Joi.string(),
+  supportLink: Joi.string(),
+  platforms: Joi.array().items(Joi.string()),
+  veteranFacingDescription: Joi.string(),
```
will be in next commit

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  primaryContact: Joi.object({
+    firstName: Joi.string().required(),
+    lastName: Joi.string().required(),
+    email: Joi.string().required(),
```
Will add this is in next commit.

reedship committed:
refactoring of schema names

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
removing dead code

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  namingConvention: Joi.string(),
+  centralizedBackendLog: Joi.string(),
+  listedOnMyHealthApplication: Joi.boolean(),
+}).options({ abortEarly: false });
```
This is one of the main issues with the image uploads and weren't included in the schema since I wasn't entirely sure what route we were going to be doing for images. GovDelivery doesn't currently allow for attachments, so we'd need to either refactor govdelivery to allow for image attachments + add the files to be added to the schema OR pass in image source urls as strings into the schema. 
The other thing we found the other day is that gmail doesn't render base64 images so I'm sure that's an issue we'd run into in other clients as well. 
The other week when talking to Brent he had given me the impression that we were going to be handling all images as source URLs for the MVP, but I'll need to double check that. To that point there still needs to be fields for the `appImage`, `patientWaitTime`, and `medicalDisclaimer` images, regardless.

reedship committed:
moved types to types dir

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
changed platforms to a string from an array

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

bradenshipley commented on routes/ProductionRequest.ts:
```diff
@@ -0,0 +1,121 @@
+  website: Joi.string(),
+  signUpLink: Joi.string(),
+  supportLink: Joi.string(),
+  platforms: Joi.array().items(Joi.string()),
```
fixed in next commit

reedship committed:
added schema field for image links

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
updated tests to account for refactor

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Template updates as requested in review:

Also added appImageLink, patientWaitTimeLink, and
medicalDisclaimerImageLink fields to support ticket.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
Merge branch 'api-4731-production-access-request-emails' of github.com:department-of-veterans-affairs/developer-portal-backend into api-4731-production-access-request-emails

reedship committed:
updated govdelivery tests to reference new production schema

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
using dot notation for organization access in govdelivery subject

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
updated consumer production template to use lists to organize steps

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

reedship committed:
updated production support template to account for new variable names.

Signed-off-by: Braden Shipley <bradenrshipley@gmail.com>

johanna-adhoc reviewed (approved):
Approved. Validation on user input will be implemented with ticket [API-8638](https://vajira.max.gov/browse/API-8638)
